source/credentials_provider_sts.c

```diff
@@ -0,0 +1,668 @@
+            goto finish;
+        }
+
+        credentials = aws_mem_calloc(provider_user_data->allocator, 1, sizeof(struct aws_credentials));
+        AWS_LOGF_DEBUG(AWS_LS_AUTH_CREDENTIALS_PROVIDER, "(credentials=%p): parsing credentials", (void *)credentials);
+        credentials->allocator = provider_user_data->allocator;
+        if (aws_xml_parser_parse(&xml_parser, s_on_node_encountered_fn, credentials)) {
+            goto finish;
+        }
+
+        aws_xml_parser_clean_up(&xml_parser);
+
+        if (!(credentials->access_key_id && credentials->secret_access_key && credentials->session_token)) {
+            AWS_LOGF_ERROR(
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                "(id=%p): credentials document was corrupted, treating as an error.",
+                (void *)provider_user_data->provider);
+            aws_credentials_destroy(credentials);
+            credentials = NULL;
+        }
+    }
+
+finish:
+    provider_user_data->callback(credentials, provider_user_data->user_data);
+
+    if (credentials) {
+        aws_credentials_destroy(credentials);
+    }
+
+    provider_impl->function_table->aws_http_connection_manager_release_connection(
+        provider_impl->connection_manager, provider_user_data->connection);
+    s_clean_up_user_data(provider_user_data);
+}
+
+/* called upon aquiring a connection from the pool */
+static void s_on_connection_setup_fn(struct aws_http_connection *connection, int error_code, void *user_data) {
+    struct sts_creds_provider_user_data *provider_user_data = user_data;
+    struct aws_credentials_provider_sts_impl *provider_impl = provider_user_data->provider->impl;
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+        "(id=%p): connection returned with error code %d",
+        (void *)provider_user_data->provider,
+        error_code);
+
+    if (error_code) {
+        aws_raise_error(error_code);
+        goto error;
+    }
+    provider_user_data->connection = connection;
+
+    if (aws_byte_buf_init(&provider_user_data->output_buf, provider_impl->provider->allocator, 2048)) {
+        provider_impl->function_table->aws_http_connection_manager_release_connection(
+            provider_impl->connection_manager, connection);
+    }
+
+    struct aws_http_make_request_options options = {
+        .manual_window_management = false,
+        .user_data = user_data,
+        .request = provider_user_data->message,
+        .self_size = sizeof(struct aws_http_make_request_options),
+        .on_response_headers = NULL,
+        .on_response_header_block_done = NULL,
+        .on_response_body = s_on_incoming_body_fn,
+        .on_complete = s_on_stream_complete_fn,
+    };
+
+    struct aws_http_stream *stream =
+        provider_impl->function_table->aws_http_connection_make_request(connection, &options);
+    if (!stream) {
+        provider_impl->function_table->aws_http_connection_manager_release_connection(
```

Comment:
Is there any way that connection release could become part of the user data cleanup so that we don't need it sprinkled around?